connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanResult.java

```diff
@@ -68,45 +71,88 @@ public PluginScanResult(
+                merge(results, PluginScanResult::predicates),
+                merge(results, PluginScanResult::configProviders),
+                merge(results, PluginScanResult::restExtensions),
+                merge(results, PluginScanResult::connectorClientConfigPolicies)
+        );
+    }
+
+    private static <T, R extends Comparable<R>> SortedSet<R> merge(List<T> results, Function<T, SortedSet<R>> accessor) {
+        SortedSet<R> merged = new TreeSet<>();
+        for (T element : results) {
+            merged.addAll(accessor.apply(element));
+        }
+        return merged;
+    }
+
+    public SortedSet<PluginDesc<SinkConnector>> sinkConnectors() {
         return sinkConnectors;
     }
 
-    public Collection<PluginDesc<SourceConnector>> sourceConnectors() {
+    public SortedSet<PluginDesc<SourceConnector>> sourceConnectors() {
         return sourceConnectors;
     }
 
-    public Collection<PluginDesc<Converter>> converters() {
+    public SortedSet<PluginDesc<Converter>> converters() {
         return converters;
     }
 
-    public Collection<PluginDesc<HeaderConverter>> headerConverters() {
+    public SortedSet<PluginDesc<HeaderConverter>> headerConverters() {
         return headerConverters;
     }
 
-    public Collection<PluginDesc<Transformation<?>>> transformations() {
+    public SortedSet<PluginDesc<Transformation<?>>> transformations() {
         return transformations;
     }
 
-    public Collection<PluginDesc<Predicate<?>>> predicates() {
+    public SortedSet<PluginDesc<Predicate<?>>> predicates() {
         return predicates;
     }
 
-    public Collection<PluginDesc<ConfigProvider>> configProviders() {
+    public SortedSet<PluginDesc<ConfigProvider>> configProviders() {
         return configProviders;
     }
 
-    public Collection<PluginDesc<ConnectRestExtension>> restExtensions() {
+    public SortedSet<PluginDesc<ConnectRestExtension>> restExtensions() {
         return restExtensions;
     }
 
-    public Collection<PluginDesc<ConnectorClientConfigOverridePolicy>> connectorClientConfigPolicies() {
+    public SortedSet<PluginDesc<ConnectorClientConfigOverridePolicy>> connectorClientConfigPolicies() {
         return connectorClientConfigPolicies;
     }
 
+    public void forEach(Consumer<PluginDesc<?>> consumer) {
+        forEach(sinkConnectors(), consumer);
+        forEach(sourceConnectors(), consumer);
+        forEach(converters(), consumer);
+        forEach(headerConverters(), consumer);
+        forEach(transformations(), consumer);
+        forEach(predicates(), consumer);
+        forEach(configProviders(), consumer);
+        forEach(restExtensions(), consumer);
+        forEach(connectorClientConfigPolicies(), consumer);
+    }
+
+    private static <T> void forEach(SortedSet<PluginDesc<T>> set, Consumer<PluginDesc<?>> consumer) {
+        for (PluginDesc<T> plugin : set) {
+            consumer.accept(plugin);
+        }
+    }
```

Comment:
This can be simplified if we tweak the type signature of `allPlugins`:
```suggestion
    public void forEach(Consumer<PluginDesc<?>> consumer) {
        allPlugins.forEach(plugins -> plugins.forEach(consumer));
    }
```